lib/rucio/core/did.py

```diff
@@ -613,19 +616,34 @@ def delete_dids(dids, account, session=None):
     # Delete rules on did
+    skip_deletion = False  # Skip deletion in case of expiration of a rule
     if rule_id_clause:
         with record_timer_block('undertaker.rules'):
-            for (rule_id, scope, name, rse_expression) in session.query(models.ReplicationRule.id,
-                                                                        models.ReplicationRule.scope,
-                                                                        models.ReplicationRule.name,
-                                                                        models.ReplicationRule.rse_expression).filter(or_(*rule_id_clause)):
+            for (rule_id, scope, name, rse_expression, locks_ok_cnt, locks_replicating_cnt, locks_stuck_cnt) in session.query(models.ReplicationRule.id,
+                                                                                                                              models.ReplicationRule.scope,
+                                                                                                                              models.ReplicationRule.name,
+                                                                                                                              models.ReplicationRule.rse_expression,
+                                                                                                                              models.ReplicationRule.locks_ok_cnt,
+                                                                                                                              models.ReplicationRule.locks_replicating_cnt,
+                                                                                                                              models.ReplicationRule.locks_stuck_cnt).filter(or_(*rule_id_clause)):
                 logging.debug('Removing rule %s for did %s:%s on RSE-Expression %s' % (str(rule_id), scope, name, rse_expression))
+
                 # Propagate purge_replicas from did to rules
                 if (scope, name) in not_purge_replicas:
                     purge_replicas = False
                 else:
                     purge_replicas = True
-                rucio.core.rule.delete_rule(rule_id=rule_id, purge_replicas=purge_replicas, delete_parent=True, nowait=True, session=session)
+                if expire_rules and locks_ok_cnt + locks_replicating_cnt + locks_stuck_cnt > 10000:
+                    # Expire the rule (soft=True)
+                    rucio.core.rule.delete_rule(rule_id=rule_id, purge_replicas=purge_replicas, soft=True, delete_parent=True, nowait=True, session=session)
+                    # Update expiration of did
+                    set_metadata(scope=scope, name=name, key='lifetime', value=3600 * 24, session=session)
+                    skip_deletion = True
+                else:
+                    rucio.core.rule.delete_rule(rule_id=rule_id, purge_replicas=purge_replicas, delete_parent=True, nowait=True, session=session)
```

Comment:
Performance wise, bulk methods will help. Do they exist, e.g.,  `delete_rules`, `bulk_set_metadata` ?
If not, then please consider storing the concerned rules in a list instead as a first step to introduce the logic.  The bulk update itself can be done with raw query instead or even better introducing the corresponding bulk methods starting with simple sequential implementation will help converging.

Reply:
No, there are no bulk methods for both of these. For `delete_rule` we should consider it, not necessarily for this workflow, but also for the judge-cleaner it would be beneficial to have a bulk method. But due to it's complexity this is on a longer timescale. 
For `set_metadata` we don't really need it (at least not for this workflow). This branch of the if statement is called so rarely that it is just extremely unlikely that there will be a second did in the same chunk where the condition applies. There are just not that many rules >10k, so it's not really worth it.